### PR TITLE
New method starknet_traceTransaction

### PR DESCRIPTION
**Pull request by kirugan**

### CODE REVIEW

**Comment by omerfirmak** [comment]
Rebase please

File: rpc/handlers.go
Comment (kirugan) [inline rpc/handlers.go:1113]: it's temporary to debug execution info

File: rpc/trace.go
Comment (omerfirmak) [inline rpc/trace.go:10]: Why are we missing the rest?

Comment (kirugan) [reply rpc/trace.go]: I've been thinking how to properly declare enum of different transaction types (choice between struct with same field + omitempty or private interface). I decided to use the latter.

File: rpc/handlers.go
Comment (omerfirmak) [inline rpc/handlers.go:1139]:
```suggestion
	state, closer, err := h.bcReader.StateAtBlockNumber(blockNumber -1)
```

File: rpc/handlers.go
Comment (omerfirmak) [inline rpc/handlers.go:1161]: This wont work. You need to query classes from the head state. Because `tx.ClassHash` wont be in the state for the block with number `blockNumber - 1` since it was declared on `blockNumber`

File: rpc/handlers.go
Comment (omerfirmak) [inline rpc/handlers.go:1166]:
```suggestion
                rpcErr := *ErrContractError
                rpcErr.Data = err.Error()
                return nil, &rpcErr
```

Let's do what EstimateFee does

File: rpc/handlers.go
Comment (omerfirmak) [inline rpc/handlers.go:1163]: You have to execute the entire block, not just the `txn`. So you need fetch the entire block with `blockByID`

File: rpc/trace.go
Comment (omerfirmak) [inline rpc/trace.go:14]: We dont need separate types for each transaction.

Let's just use something like

```
type TransactionTrace struct {
	ValidateInvocation    *FunctionInvocation `json:"validate_invocation,omitempty"`
	ExecuteInvocation     *ExecuteInvocation  `json:"execute_invocation,omitempty"`
	FeeTransferInvocation *FunctionInvocation `json:"fee_transfer_invocation,omitempty"`
}
```

and some of the fields can be `nil` according to the transaction type

**Review by omerfirmak — CHANGES_REQUESTED** [review]

File: rpc/handlers.go
Comment (kirugan) [reply rpc/handlers.go:1159]: Agree sounds reasonable, but why does it work with current implementation:
```
curl --location --request POST 'http://localhost:6060' \
--header 'Content-Type: application/json' \
--data-raw '{
    "jsonrpc": "2.0",
    "id": 1,
    "method": "starknet_traceTransaction",
    "params": {
        "transaction_hash": "0x55ba350e706f02c5903cc62f2381d6a7cc6f88efeb9c72a230487fec5f60a50"
    }
}'
```
on commit 7de8ce6a9c8bfeafe04fa5152599a04257382eca, I logged locally `append(classes, class.Class)` logic - it seems working, maybe not the way expected 😄

Comment (kirugan) [reply rpc/handlers.go:1159]: Probably it worked because class is declared at 4883 block and tx is from 4968

File: rpc/handlers.go
Comment (kirugan) [reply rpc/handlers.go]: Fixed, but now it doesn't work :) the problem that if block contains transaction type that unsupported by `vm.marshalTxn(...)`  then entire execution is failed. Could you please explain why do I need entire block to execute? because a transaction that I requested can rely on another in the same block?

File: rpc/handlers.go
Comment (kirugan) [inline rpc/handlers.go]: @omerfirmak what if there is a request for tx that resides in 0 block? what should we do in this case

File: rpc/handlers.go
Comment (omerfirmak) [reply rpc/handlers.go]: >  Probably it worked because class is declared at 4883 block and tx is from 4968

Yeap, indeed. It is just redeclaring a class hash. Which is fine.

File: rpc/handlers.go
Comment (omerfirmak) [reply rpc/handlers.go]: > because a transaction that I requested can rely on another in the same block?

Yes, exactly.

> if block contains transaction type that unsupported by vm.marshalTxn(...)

Let's add support for them then:)

File: rpc/handlers.go
Comment (omerfirmak) [reply rpc/handlers.go]: Yes, that is an issue. Let's add a special case in `StateAtBlockNumber`. 

```
var EmptyState := core.NewState(db.NewMemTransaction)

if blockNumber == 0 {
   return EmptyState
}
```

Comment (omerfirmak) [reply rpc/handlers.go]: and in trace handler, we can just use `h.bcReader.StateAtBlockHash(block.ParentHash)` instead of `h.bcReader.StateAtBlockNumber(blockNumber -1)`

File: rpc/handlers.go
Comment (omerfirmak) [inline rpc/handlers.go:1158]: We need to check for pending block explicitly, since header.Number is not valid for pending blocks. See how we do that in `EstimateFee`

Comment (kirugan) [reply rpc/handlers.go]: Added support

File: rpc/handlers.go
Comment (kirugan) [reply rpc/handlers.go]: Added

File: rpc/trace.go
Comment (kirugan) [reply rpc/trace.go]: Changed type from interface to struct

File: rpc/handlers.go
Comment (kirugan) [reply rpc/handlers.go]: It was easier than I expected 😄

File: rpc/trace.go
Comment (omerfirmak) [reply rpc/trace.go]: Can we make them pointers as well? So it is easier to check if they are empty or no

File: grpc/gen/kv.pb.go
Comment (omerfirmak) [inline grpc/gen/kv.pb.go:1]: can you remove these changes in `grpc` and `l1` since they are not-related

File: rpc/trace.go
Comment (omerfirmak) [inline rpc/trace.go:39]: is this a directive to compiler or just a regular comment?

Comment (kirugan) [reply rpc/trace.go:39]: Just a regular comment

File: grpc/gen/kv.pb.go
Comment (kirugan) [reply grpc/gen/kv.pb.go:1]: Sure, there were some issues with file watcher in GoLand IDE. Fixed

File: rpc/trace.go
Comment (kirugan) [reply rpc/trace.go]: Agree, fixed

File: vm/vm.go
Comment (omerfirmak) [inline vm/vm.go:166]: can we move this to caller site? it feels like it doesnt belong here

File: vm/rust/src/lib.rs
Comment (omerfirmak) [inline vm/rust/src/lib.rs:244]: Let's not copy and paste `Transaction::from_api` here. 

Instead let's check for the cases that `Transaction::from_api` panic and make sure we error out early don't call it in those cases.

File: rpc/trace.go
Comment (omerfirmak) [inline rpc/trace.go:14]: Do you think we can just use `vm` Types for this?
This looks way too much code to duplicate

Comment (kirugan) [reply rpc/trace.go:14]: Unfortunately no, because `CallInfo` has a structure that depends on format of rust-vm (and naming of json fields) and `FunctionInvocation` type is format of json-rpc. For instance:
```
result.Result = info.Execution.Retdata
```
json rpc field `result` is actually `retdata` inside other struct in vm response

Comment (omerfirmak) [reply rpc/trace.go:14]: Can we fix stuff like this in rust types?

For example instead of 

``` rust
#[derive(Debug,Serialize)]
pub struct CallExecution {
    pub retdata: Retdata,
    pub events: Vec<OrderedEvent>,
    pub l2_to_l1_messages: Vec<OrderedL2ToL1Message>,
}
```

can we rename `retdata` to `result` here?

We have 3 almost-exactly same type definitions, 1 in Rust, 2 in Go. If we can at least get rid of one of them. It would be already an huge improvement.

File: vm/vm.go
Comment (kirugan) [reply vm/vm.go]: Moved

File: rpc/trace.go
Comment (omerfirmak) [reply rpc/trace.go:14]: I wouldnt mind even if we straight up used the JSON that we return from rust. Without unmarshaling/marshaling on Go side.

File: vm/rust/src/lib.rs
Comment (kirugan) [reply vm/rust/src/lib.rs:223]: Ok I added call to `Transaction::from_api`, also decided to make things clean to keep validation in existing function `transaction_from_api(...)`

File: rpc/handlers.go
Comment (omerfirmak) [inline rpc/handlers.go:1173]: nit

```suggestion
	sequencerAddress := header.SequencerAddress
	if sequencerAddress == nil {
		sequencerAddress = core.NetworkBlockHashMetaInfo(h.network).FallBackSequencerAddress
	}
```

Comment (kirugan) [reply rpc/handlers.go]: Thank you, it's much cleaner now

File: rpc/handlers.go
Comment (omerfirmak) [inline rpc/handlers.go:1203]: this got reverted for some reason, can you make sure nothing else is mistakenly reverted as well?

File: vm/rust/src/lib.rs
Comment (jelilat) [inline vm/rust/src/lib.rs:229]: This line is causing the following error even in cases where the transaction is not a deploy transaction. 
```
"error":{"code":40,"message":"Contract error","data":"Deploy transaction is not supported 
```

Looking at it, I realised that the error shows up in blocks that contain deploy transactions even though the transaction in question is not a Deploy transaction. 

Do we need to go through the whole block to get the transaction trace for a single transaction?

Comment (omerfirmak) [reply vm/rust/src/lib.rs:229]: > Do we need to go through the whole block to get the transaction trace for a single transaction?

Yes, because transactions get effected from state generated by the previous transactions in the same block. For trace to be accurate we need to execute the entire block

Comment (jelilat) [reply vm/rust/src/lib.rs:229]: Okay

File: rpc/handlers.go
Comment (omerfirmak) [inline rpc/handlers.go:1166]: should be a ErrTxnHashNotFound

File: rpc/handlers.go
Comment (omerfirmak) [inline rpc/handlers.go:1183]: One thing I didnt think of before,

If txn `hash` is part of the pending block, we need to use `PendingState` here, instead of `HeadState`

File: vm/vm.go
Comment (omerfirmak) [inline vm/vm.go:53]:
```suggestion
	traces      []json.RawMessage
```

so we dont need to convert it when returning from `vm.Execute`?


``` golang
	var traces []json.RawMessage
	for _, trace := range context.traces {
		traces = append(traces, json.RawMessage(trace))
	}
```

File: utils/slices.go
Comment (omerfirmak) [inline utils/slices.go:33]: do we still need these?

File: vm/rust/src/lib.rs
Comment (omerfirmak) [inline vm/rust/src/lib.rs:246]: leftover?

File: utils/slices.go
Comment (kirugan) [reply utils/slices.go:33]: I left these helpers for future use. They quite helpful

File: rpc/handlers.go
Comment (joshklop) [inline rpc/handlers.go:1236]: Do we need to re-execute the whole block, or just `block.Transactions[:txIndex]`?

Comment (omerfirmak) [reply rpc/handlers.go:1236]: > or just block.Transactions[:txIndex]?

Yeap, nice catch.

File: rpc/handlers.go
Comment (kirugan) [reply rpc/handlers.go]: yep, thanks

File: rpc/handlers.go
Comment (joshklop) [reply rpc/handlers.go:1239]: That would make sense to me. If class isn't present in the state before `txIndex`, we shouldn't need to include it.

Comment (kirugan) [reply rpc/handlers.go:1240]: > we shouldn't need to include it.

added

**Review by omerfirmak — APPROVED** [review]
LGTM, let's wait for @joshklop's approval as well.

File: blockchain/blockchain.go
Comment (joshklop) [inline blockchain/blockchain.go:718]: Maybe the caller should create the empty state?

**Review by joshklop — APPROVED** [review]
The code looks good to me. Have we done manual tests for this yet?

File: blockchain/blockchain.go
Comment (kirugan) [reply blockchain/blockchain.go:718]: Hm.. not sure, we would implement it over and over again on caller side I guess. @omerfirmak wdyt?

Comment (omerfirmak) [reply blockchain/blockchain.go:718]: > the caller should create the empty state

I think it fits well in the API of `blockchain.Reader`, we can keep it here.
We can make `emptyState` a singleton tho, to avoid instantiating it every time. But it's not a priority either.

Comment (kirugan) [reply blockchain/blockchain.go:718]: > We can make emptyState a singleton tho, to avoid instantiating it every time. But it's not a priority either.

Good point I'll add it to next PR

**Pull request merged**
